### ISSUE TITLE
thread 'rustc' panicked at 'Trying to convert unsized value to lval'

### ISSUE THREAD

**Issue by jarrett** [rust-lang/rust#21847]
I got this compiling a fairly large app:

    error: internal compiler error: unexpected panic
    note: the compiler unexpectedly panicked. this is a bug.
    note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
    note: run with `RUST_BACKTRACE=1` for a backtrace
    thread 'rustc' panicked at 'Trying to convert unsized value to lval', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/librustc_trans/trans/datum.rs:402

    stack backtrace:
       1:        0x10c85ecd7 - sys::backtrace::write::h4ee5c5ad264b65f5mbu
       2:        0x10c884e9c - failure::on_fail::h55e5ff39107f11fcHbB
       3:        0x10c7e6618 - rt::unwind::begin_unwind_inner::habba3c5d37f59591DTA
       4:        0x10904eca5 - rt::unwind::begin_unwind::h5615917778268573803
       5:        0x1090d753e - trans::datum::Datum<'tcx, Expr>::to_lvalue_datum::h9595aad7c2992d4cZSf
       6:        0x1090feaac - trans::expr::trans_index::h33588fb9dd7ce121yzi
       7:        0x1090e75a3 - trans::expr::trans_unadjusted::h4bfe4622405612c0Vki
       8:        0x10909f7d4 - trans::expr::trans::h87900629b98dec94YCh
       9:        0x109101448 - trans::expr::trans_addr_of::he4d93cb9897bf566uHj
      10:        0x1090e779e - trans::expr::trans_unadjusted::h4bfe4622405612c0Vki
      11:        0x10909f7d4 - trans::expr::trans::h87900629b98dec94YCh
      12:        0x10909e521 - trans::expr::trans_into::h8d099f5179d39277uzh
      13:        0x10911d773 - trans::expr::trans_adt::h04652ea6bac31317Drj
      14:        0x1090b0cdf - trans::base::trans_named_tuple_constructor::h27bf054814ac4386Dcu
      15:        0x1090e25a6 - trans::callee::trans_call_inner::h12337891083913942824
      16:        0x1090e9b77 - trans::expr::trans_rvalue_dps_unadjusted::h76d104b0c393ded0ITi
      17:        0x10909e762 - trans::expr::trans_into::h8d099f5179d39277uzh
      18:        0x10909f339 - trans::controlflow::trans_block::ha4c3bc4d960bd55aT3d
      19:        0x1090e8c91 - trans::expr::trans_rvalue_dps_unadjusted::h76d104b0c393ded0ITi
      20:        0x10909e762 - trans::expr::trans_into::h8d099f5179d39277uzh
      21:        0x10909f339 - trans::controlflow::trans_block::ha4c3bc4d960bd55aT3d
      22:        0x10916c644 - trans::base::trans_closure::hc2f7f2fc0f6c7df0WXt
      23:        0x10908b1c9 - trans::base::trans_fn::he8bb083f0eb63cd3Q8t
      24:        0x109086f0b - trans::base::trans_item::h6e886ccbd5f66d227vu
      25:        0x10916de29 - trans::base::trans_mod::hacffe7d798632c0cIBu
      26:        0x10908664e - trans::base::trans_item::h6e886ccbd5f66d227vu
      27:        0x109172a8c - trans::base::trans_crate::h03f93f510f95d199hrv
      28:        0x108f29f6e - driver::phase_4_translate_to_llvm::hf01e1f3e6069eac9mOa
      29:        0x108f05f3c - driver::compile_input::h8e0204b8bb9ea44bBba
      30:        0x108fced3e - run_compiler::h322f9334a9a1d7e6l9b
      31:        0x108fcbd7f - thunk::F.Invoke<A, R>::invoke::h14356237310415989951
      32:        0x108fcaa10 - rt::unwind::try::try_fn::h798029436104877783
      33:        0x10c8fb329 - rust_try_inner
      34:        0x10c8fb316 - rust_try
      35:        0x108fcb0d4 - thunk::F.Invoke<A, R>::invoke::h15628186897385384602
      36:        0x10c86f553 - sys::thread::thread_start::h3f48e8c893476626A3w
      37:     0x7fff854c5899 - _pthread_body
      38:     0x7fff854c572a - _pthread_struct_init

The problem appears to be in this function:

    pub fn chunk_containing(&self, abs_x: u32, abs_y: u32) -> Option<&Chunk> {
        // x_idx and y_idx are indices int the chunks array.
        let x_idx: usize = abs_x.div_floor(&self.chunk_x_size) as usize;
        let y_idx: usize = abs_y.div_floor(&self.chunk_y_size) as usize;
        
        if ((x_idx as u32) < self.x_chunks && (y_idx as u32) < self.y_chunks) {
            Some(&self.chunks[y_idx][x_idx])
        } else {
            None
        }
    }

And specifically, the problem appears to be this line:

    Some(&self.chunks[y_idx][x_idx])

I reached that conclusion because replacing the above line with `None` clears the error. The type of `self.chunks` is `[[Chunk]]`. Changing its type to `Vec<Vec<Chunk>>` also clears the error.

OSX 10.9.5
rustc 1.0.0-nightly (1d00c545e 2015-01-30 19:56:34 +0000)
binary: rustc
commit-hash: 1d00c545ede609b9d43fdf9f252c15da5a66dac7
commit-date: 2015-01-30 19:56:34 +0000
host: x86_64-apple-darwin
release: 1.0.0-nightly

**Comment by reima** [comment]
I'm also affected by this bug. Here's a minimal example which produces the ICE:

```rust
const NUMS: &'static [u32] = &[1, 2, 3];

fn main() {
  let x = 0;
  NUMS[x];
}
```

This is the backtrace I get:
```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Trying to convert unsized value to lval', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc_trans/trans/datum.rs:495

stack backtrace:
   1:     0x7fcb21cedb70 - sys::backtrace::write::h00fa41ac57b8527bfmB
   2:     0x7fcb21d164f0 - panicking::on_panic::he60a0e1a593653a5TOK
   3:     0x7fcb21c5d960 - rt::unwind::begin_unwind_inner::h3bcb1406fb794dc7etK
   4:     0x7fcb208ed3e0 - rt::unwind::begin_unwind::h14051137956471406385
   5:     0x7fcb209661d0 - trans::datum::Datum<'tcx, Expr>::to_lvalue_datum::hccf6a8b7b600451ebYf
   6:     0x7fcb20998fe0 - trans::expr::trans_index::h0e1e4b8610393db20Ei
   7:     0x7fcb20985bc0 - trans::expr::trans_unadjusted::hfa1c05df16b707e2mqi
   8:     0x7fcb2093acd0 - trans::expr::trans_into::h546d5b1120c5d84ejGh
   9:     0x7fcb2093a1a0 - trans::controlflow::trans_stmt_semi::h4b02d82dbafc77ce6de
  10:     0x7fcb2093b840 - trans::controlflow::trans_block::h231f411ce6a4283aXee
  11:     0x7fcb20a07b80 - trans::base::trans_closure::hd62fcc69afa5e29aFiu
  12:     0x7fcb209271a0 - trans::base::trans_fn::h58b20f0749af6eafWtu
  13:     0x7fcb20922810 - trans::base::trans_item::h6046484f0ec08588PSu
  14:     0x7fcb20a0f0e0 - trans::base::trans_crate::h7924470e1c25a280fQv
  15:     0x7fcb22358cb0 - driver::phase_4_translate_to_llvm::hb5ebc8e27de986a12Oa
  16:     0x7fcb22332070 - driver::compile_input::h5ed7326e88136f7fEba
  17:     0x7fcb22402290 - run_compiler::h12bed89c4496bc43Bbc
  18:     0x7fcb224007f0 - thunk::F.Invoke<A, R>::invoke::h13875370646425574429
  19:     0x7fcb223ff6d0 - rt::unwind::try::try_fn::h5086316350382225766
  20:     0x7fcb21d83050 - rust_try_inner
  21:     0x7fcb21d83040 - rust_try
  22:     0x7fcb223ff980 - thunk::F.Invoke<A, R>::invoke::h12662570625570879817
  23:     0x7fcb21d01e80 - sys::thread::thread_start::h197b42630d52606aYVF
  24:     0x7fcb1be0f0c0 - start_thread
  25:     0x7fcb218d6fd9 - __clone
  26:                0x0 - <unknown>
```

Environment:

Ubuntu 14.04.2 LTS
Linux 3.13.0-45-generic #74-Ubuntu SMP Tue Jan 13 19:36:28 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

rustc 1.0.0-nightly (81bce5290 2015-02-16) (built 2015-02-17)
binary: rustc
commit-hash: 81bce5290ff55b9a2eddd83d31b0778180904d7f
commit-date: 2015-02-16
build-date: 2015-02-17
host: x86_64-unknown-linux-gnu
release: 1.0.0-nightly

**Comment by SimonSapin** [comment]
This minimal example given above is fixed in rustc 1.0.0-nightly (4db0b3246 2015-02-25) (built 2015-02-26).

**Comment by steveklabnik** [comment]
Yes, the minimal example does not for me, either. @jarrett if this still occurs in your big app, let me know.
